ls1intum/Artemis
# Feature/Programming Exercise/Sequential test runs

### Checklist


### Motivation and Context
Atm structural and behavior tests are executed in the same build job.

However until structural tests are successful, students should concentrate only on them.
Because of this, we should first run structural and then - only if the structural tests are successful - also the behavior tests.

### Description

- There is now an option (checkbox) in programming exercises to create an exercise with sequential test execution
- Sequential means that first structural tests are executed, then behavioral tests
- The new option changes
1. How the build plans are generated: one job for executing tests vs two jobs
2. That we now have to persist test cases because otherwise we can't interpret an incoming result (see below)
3. How the results of a build are interpreted: if now a test's results are missing from the results array, it means that it was not executed
4. How the score is calculated: We now have to recalculate the score of an incoming result as it might not include all test cases. This already also calculates the weight based on the test cases weights
5. How we show the results in the client (see also screenshots below): There are now the following TestCaseStates -> SUCCESS (positive result), FAIL (negative result), NOT_EXECUTED (no result in results array, but results array exists), NO_RESULT (no results array exists, build failed)
- Test cases are extracted on the server from the solution build result (as this is the only one that now contains all test results)
1. I chose this approach because it is language agnostic
2. Test cases will not be deleted if they are not anymore part of the solution result, but set to inactive. This is done as a preparation for the solution diff (outdated markers). 

Problems open to solved in the implementation:
- [x] Solution build run must be triggered before the template build run, because it is used to extract the test cases, and if it has not completed yet, the template build run will calculate the wrong number of tests
- [x] Handle case of no exercise id in programming exercise creation dialog
- [x] When there are results, but not for all tests, AND they are not part of the feedback list, the feedback modal can still be opened (client bug)
- [x] ~Within the interactive instructions, it should also not only say which tests fail, but which were not ran~ -> this seemed to make sense, but did not feel good to me.
- [x] Write client tests
- [x] Write server tests
- [x] How do we validate test cases in the student instructions?
- [x] Failed build result should not be altered (server bug)
- [x] ~Not sure if related, but deletion of a sequential exercise failed (api/programming-exercises/551?deleteStudentReposBuildPlans=true&deleteBaseReposBuildPlans=true -> 500)~ could not reproduce
- [x] Non sequential exercise does not have test cases (seemed to work after executing another build)
- [x] 3 not executed -> remove from result string, move information into result details
- [x] Prepend test name with class name (or not? only if duplicate?)
=> I think we should count this as a new feature request. In this aspect the new system would behave similar to the old one: Before saving the test cases, I store them in a `Set`. So the last found test case with the same name would win. There also won't be a unique constraint failed error in runtime.
- [x] Test sequential test runs & surfire report locally
=> The surfire-reports folder is cleared when the second test command is ran. I think that Bamboo makes sure to read the reports from every task and cummulate them. We could look further into it, but it doesn't seem to have any effects on the build output.
- [x] Remove trigger from solution to template, add manual trigger button 
- [x] Also in instructions markup: 0 of X passed instead of 0 of failed.
- [x] We don't need a build plan for the test repo, this is done by the bitbucket trigger
- [x] How to distinguish between structural & behavior tests in Java? @krusche can help.
=> test repository: https://github.com/thilo-behnke/test-example
- [x] If sequential test runs are active for java, change the folder structure (as discussed with @krusche)

### Steps for Testing


2 scenarios:

a) New exercise

1. Create a new programming exercise
2. As long as structural tests fail, behavior tests should not be executed

b) Old exercise

Nothing should change.

### Screenshots
![image](https://user-images.githubusercontent.com/28230611/60400425-c7f86580-9b73-11e9-9a51-37df4ca48a6b.png)

![image](https://user-images.githubusercontent.com/28230611/60400431-dba3cc00-9b73-11e9-9f0e-95065b11dfad.png)
